src/protocols/rdp/channels/rdpsnd/rdpsnd.c

```diff
@@ -35,11 +35,23 @@ void guac_rdpsnd_process_receive(guac_rdp_common_svc* svc,
     guac_rdpsnd* rdpsnd = (guac_rdpsnd*) svc->data;
     guac_rdpsnd_pdu_header header;
 
+    /* Check that we at least have a header. */
+    if (Stream_GetRemainingLength(input_stream) < 4)
+        return;
+    
     /* Read RDPSND PDU header */
     Stream_Read_UINT8(input_stream, header.message_type);
     Stream_Seek_UINT8(input_stream);
     Stream_Read_UINT16(input_stream, header.body_size);
-
+    
+    if (Stream_GetRemainingLength(input_stream) < header.body_size) {
+        guac_client_log(svc->client, GUAC_LOG_DEBUG, "Not enough bytes in stream."
+                "  Remaining: %d, Body size: %d",
+                Stream_GetRemainingLength(input_stream),
+                header.body_size);
+        return;
+    }
```

Comment:
This area here is causing me a little trouble - for whatever reason, if I do this check, sound doesn't work.  Maybe I'm doing some pointer access incorrectly or something??

Reply:
I'm not a RDP expert. It's just a guess, based on reading the comments in the source code.
It looks like some PDUs like do not have headers, so there is little sense in inspecting header.body_size for them...
Maybe you need to move this check a few lines lower (i.e. put it after the following piece of code)?
https://github.com/apache/guacamole-server/blob/9c37fc56171d55b8fa838949d7fd40db85506857/src/protocols/rdp/channels/rdpsnd/rdpsnd.c#L43-L50

Reply:
Thanks for the hint @trengri - I will take a look!

Reply:
Actually, looking further at the code, it looks like each of the functions that are called in the switch statement already do the Stream size verification prior to reading, so maybe there's no need to do it, here??
* `guac_rdpsnd_wave_handler()` is doing Stream write operations, not reads.
* `guac_rdpsnd_formats_handler`, `guac_rdpsnd_training_handler`, and `guac_rdpsnd_wave_info_handler` already do the check.
* `guac_rdpsnd_close_handler` doesn't do any reads (or anything else).

So, seems like checking within this function may not be necessary?

Reply:
So, went ahead and moved it down below that first check and it seems to be working fine.  Still not sure it's strictly necessary since we're checking it in the other functions, but works for now!